src/storm/modelchecker/helper/indefinitehorizon/visitingtimes/SparseDeterministicVisitingTimesHelper.h

```diff
@@ -100,30 +116,53 @@ class SparseDeterministicVisitingTimesHelper : public SingleValueModelCheckerHel
 
     /*!
-     * @return the environment used when solving non-trivial SCCs.
+     * @post _nonBsccStates points to the vector of non-BSCC states
+     */
+    void createNonBsccStateVector();
+
+    /*!
+     * Computes for each state an upper bound on the expected number of times we are visiting that state.
+     * @note The upper bounds are computed using techniques from by Baier et al. [CAV'17] (https://doi.org/10.1007/978-3-319-63387-9_8)
+     * @param stateSetAsBitVector the states for which the upper bounds are computed.
+     * @return for each state the upper bound on the expected number of times that state is visited.
      */
-    storm::Environment getEnvironmentForSccSolver(storm::Environment const& env) const;
+    std::vector<ValueType> computeUpperBounds(storm::storage::BitVector const& stateSetAsBitVector) const;
+
+    /*!
+     * Adapts the precision of the solving method if necessary (i.e., if the model is a CTMC or when a topological solving method is used).
+     * @param env The environment, containing information on the precision that should be achieved.
+     * @param topological If set to true, the environment when solving non-trivial SCCs for topological solving method is returned.
+     * @return the environment used when solving the equation system(s) for EVTs with adapted precision for a topological solving method or CTMCs.
+     */
+    storm::Environment getEnvironmentForSolver(storm::Environment const& env, bool topological = false) const;
+
+    /*!
+     * @return the environment used when solving non-trivial SCCs for topological solving method.
+     */
+    storm::Environment getEnvironmentForTopologicalSolver(storm::Environment const& env) const;
 
     /*!
      * Processes (bottom or non-bottom SCCs consisting of a single state). The resulting value is directly inserted into stateValues
      */
     void processSingletonScc(uint64_t sccState, std::vector<ValueType>& stateValues) const;
 
     /*!
-     * Solves the equation system for non-trivial SCCs (i.e. non-bottom SCCs with more than 1 state).
-     * @return for each state of the given SCC the expected number of times that state is visited.
+     * Solves the equation system for non-singleton (subs)sets of the chain's non-bottom states.
+     * @return for each state of the given set the expected number of times that state is visited.
      */
-    std::vector<ValueType> computeValueForNonTrivialScc(storm::Environment const& env, storm::storage::BitVector const& sccAsBitVector,
-                                                        std::vector<ValueType> const& stateValues) const;
+    std::vector<ValueType> computeValueForStateSet(storm::Environment const& env, storm::storage::BitVector const& stateSetAsBitVector,
+                                                   std::vector<ValueType> const& stateValues) const;
+
+    storm::storage::SparseMatrix<ValueType> const& transitionMatrix;
+    std::vector<ValueType> const* exitRates;
```

Comment:
are these const* intended?

Reply:
Thanks, this is a case for `storm::OptionalRef`